controllers/dscinitialization/utils.go

```diff
@@ -18,132 +21,178 @@ import (
-				if err != nil && !k8serr.IsAlreadyExists(err) {
-					log.Error(err, "Unable to create namespace", "name", monitoringName)
-					return err
-				}
-			} else {
-				log.Error(err, "Unable to fetch monitoring namespace", "name", monitoringName)
-				return err
-			}
-		} else { // force to patch monitoring namespace with label for cluster-monitoring
-			log.Info("Patching monitoring namespace", "name", monitoringName)
-			labelPatch := `{"metadata":{"labels":{"openshift.io/cluster-monitoring":"true", "pod-security.kubernetes.io/enforce":"baseline","opendatahub.io/generated-namespace": "true"}}}`
-
-			err = r.Patch(ctx, foundMonitoringNamespace, client.RawPatch(types.MergePatchType, []byte(labelPatch)))
-			if err != nil {
-				return err
-			}
-		}
+	// Patch monitoring namespace
+	err := r.patchMonitoringNS(ctx, dscInit, log)
+	if err != nil {
+		log.Error(err, "error patch monitoring namespace for managed cluster")
+		return err
 	}
 
 	// Create default NetworkPolicy for the namespace
-	err = r.reconcileDefaultNetworkPolicy(ctx, name, dscInit, platform)
+	err = r.reconcileDefaultNetworkPolicy(ctx, nsName, dscInit, platform, log)
 	if err != nil {
-		log.Error(err, "error reconciling network policy ", "name", name)
+		log.Error(err, "error reconciling network policy ", "name", nsName)
 		return err
 	}
 
 	// Create odh-common-config Configmap for the Namespace
-	err = r.createOdhCommonConfigMap(ctx, name, dscInit)
+	err = r.createOdhCommonConfigMap(ctx, nsName, dscInit, log)
 	if err != nil {
 		log.Error(err, "error creating configmap", "name", "odh-common-config")
 		return err
 	}
 
 	// Create default Rolebinding for the namespace
-	err = r.createDefaultRoleBinding(ctx, name, dscInit)
+	err = r.createDefaultRoleBinding(ctx, nsName, dscInit, log)
 	if err != nil {
-		log.Error(err, "error creating rolebinding", "name", name)
+		log.Error(err, "error creating rolebinding", "name", nsName)
 		return err
 	}
 	return nil
 }
 
-func (r *DSCInitializationReconciler) createDefaultRoleBinding(ctx context.Context, name string, dscInit *dsciv1.DSCInitialization) error {
-	log := logf.FromContext(ctx)
+func (r *DSCInitializationReconciler) appNamespaceHandler(ctx context.Context, nsName string, log logr.Logger) error {
+	// Check if application namespace has label "opendatahub.io/application-namespace:true"
+	// if no such namespace exist, we create it with generated-namespace and security label
+	// if namespace exist but no label, we exit
+	// if namespace exist and has label, we enforce security label onto it.
+	appNamespace := &corev1.Namespace{}
+	desiredAppNS := &corev1.Namespace{
+		ObjectMeta: metav1.ObjectMeta{
+			Name: nsName,
+		},
+	}
+	if err := r.Get(ctx, client.ObjectKeyFromObject(desiredAppNS), appNamespace); err != nil {
+		if !k8serr.IsNotFound(err) {
+			return err
+		}
+		return r.createAppNamespace(ctx, nsName)
+	}
+
+	l := appNamespace.GetLabels()
+	if l[labels.CustomizedAppNamespace] == "true" {
```

Comment:
since you are using `CreateOrUpdate` in `createAppNamespace`, can't we do this check as part of the `CreateOrUpdate` mutate function ? ?